// derby/lib/PathMap.js
// Generated by CoffeeScript 1.3.1
var PathMap,
  __slice = [].slice;

module.exports = PathMap = function() {
  this.clear();
};

PathMap.prototype = {
  clear: function() {
    this.count = 0;
    this.ids = {};
    this.paths = {};
    this.arrays = {};
  },
  id: function(path) {
    var id;
    return this.ids[path] || (this.paths[id = ++this.count] = path, this._indexArray(path, id), this.ids[path] = id);
  },
  _indexArray: function(path, id) {
    var arr, index, match, nested, remainder, set, setArrays;
    while (match = /^(.+)\.(\d+)(\*?(?:\..+|$))/.exec(path)) {
      path = match[1];
      index = +match[2];
      remainder = match[3];
      arr = this.arrays[path] || (this.arrays[path] = []);
      set = arr[index] || (arr[index] = {});
      if (nested) {
        setArrays = set.arrays || (set.arrays = {});
        setArrays[remainder] = true;
      } else {
        set[id] = remainder;
      }
      nested = true;
    }
  },
  _incrItems: function(path, map, start, end, byNum, oldArrays, oldPath) {
    var arrayMap, arrayPath, arrayPathTo, i, id, ids, itemPath, remainder, _i;
    if (oldArrays == null) {
      oldArrays = {};
    }
    for (i = _i = start; start <= end ? _i < end : _i > end; i = start <= end ? ++_i : --_i) {
      if (!(ids = map[i])) {
        continue;
      }
      for (id in ids) {
        remainder = ids[id];
        if (id === 'arrays') {
          for (remainder in ids[id]) {
            arrayPath = (oldPath || path) + '.' + i + remainder;
            if (arrayMap = oldArrays[arrayPath] || this.arrays[arrayPath]) {
              arrayPathTo = path + '.' + (i + byNum) + remainder;
              this.arrays[arrayPathTo] = arrayMap;
              this._incrItems(arrayPathTo, arrayMap, 0, arrayMap.length, 0, oldArrays, arrayPath);
            }
          }
          continue;
        }
        itemPath = path + '.' + (i + byNum) + remainder;
        this.paths[id] = itemPath;
        this.ids[itemPath] = +id;
      }
    }
  },
  _delItems: function(path, map, start, end, len, oldArrays) {
    var arrayLen, arrayMap, arrayPath, i, id, ids, itemPath, remainder, _i;
    if (oldArrays == null) {
      oldArrays = {};
    }
    for (i = _i = start; start <= len ? _i < len : _i > len; i = start <= len ? ++_i : --_i) {
      if (!(ids = map[i])) {
        continue;
      }
      for (id in ids) {
        if (id === 'arrays') {
          for (remainder in ids[id]) {
            arrayPath = path + '.' + i + remainder;
            if (arrayMap = this.arrays[arrayPath]) {
              arrayLen = arrayMap.length;
              this._delItems(arrayPath, arrayMap, 0, arrayLen, arrayLen, oldArrays);
              oldArrays[arrayPath] = arrayMap;
              delete this.arrays[arrayPath];
            }
          }
          continue;
        }
        itemPath = this.paths[id];
        delete this.ids[itemPath];
        if (i > end) {
          continue;
        }
        delete this.paths[id];
      }
    }
    return oldArrays;
  },
  onRemove: function(path, start, howMany) {
    var end, len, map, oldArrays;
    if (!(map = this.arrays[path])) {
      return;
    }
    end = start + howMany;
    len = map.length;
    oldArrays = this._delItems(path, map, start, end + 1, len);
    this._incrItems(path, map, end, len, -howMany, oldArrays);
    map.splice(start, howMany);
  },
  onInsert: function(path, start, howMany) {
    var end, len, map, oldArrays;
    if (!(map = this.arrays[path])) {
      return;
    }
    end = start + howMany;
    len = map.length;
    oldArrays = this._delItems(path, map, start, end + 1, len);
    this._incrItems(path, map, start, len, howMany, oldArrays);
    while (howMany--) {
      map.splice(start, 0, {});
    }
  },
  onMove: function(path, from, to, howMany) {
    var afterFrom, afterTo, items, map, oldArrays;
    if (!(map = this.arrays[path])) {
      return;
    }
    afterFrom = from + howMany;
    afterTo = to + howMany;
    if (from > to) {
      oldArrays = this._delItems(path, map, to, afterFrom, afterFrom);
      this._incrItems(path, map, to, from, howMany, oldArrays);
    } else {
      oldArrays = this._delItems(path, map, from, afterTo, afterTo);
      this._incrItems(path, map, afterFrom, afterTo, -howMany, oldArrays);
    }
    this._incrItems(path, map, from, afterFrom, to - from, oldArrays);
    items = map.splice(from, howMany);
    map.splice.apply(map, [to, 0].concat(__slice.call(items)));
  }
};
